jalantechnologies/boilerplate-mern
# Prashant/feature/share task

# Description

This PR adds the functionality to share tasks with multiple users within the TODO application. The following changes have been made:

- Implemented an endpoint to share tasks: `POST /api/tasks/:id/share`.
- Created a database schema for shared tasks.
- Developed unit tests to verify the sharing functionality.
- Verified the implementation using Postman for manual testing.

# Why is this change needed?

The task-sharing feature allows users to share tasks with other users on the platform, enhancing collaboration and task management capabilities.

# API Changes

## New Endpoint

**POST** `/api/tasks/:id/share`

### Request Body
- `userIds`: array of user IDs with whom the task will be shared

### Response
- `message`: confirmation message indicating successful sharing of the task

# Database Changes

A new `shared-tasks` collection has been added with the following fields:
- `taskId`: ID of the task being shared
- `userId`: ID of the user with whom the task is shared

# Tests

## Automated Test Cases Added:

**Test 1:** Verify that a task can be shared with multiple users.
- Create a task and share it with multiple users.
- Verify the task is shared correctly.

**Test 2:** Ensure an error is returned if the task does not exist.
- Attempt to share a non-existent task.
- Verify an appropriate error message is returned.

**Test 3:** Ensure tasks marked as deleted are not returned.
- Mark a task as deleted and verify it is not returned in the task list.

**Test 4:** Verify that tasks not belonging to the logged-in account are not returned.
- Create tasks for different accounts and ensure only the logged-in account's tasks are returned.

## Description of Automated Test 1

- Test for successful sharing of a task.
- Create a task and share it with multiple users.
- Verify the task is shared correctly.

## Description of Automated Test 2

- Test for error when the task does not exist.
- Attempt to share a non-existent task.
- Verify an appropriate error message is returned.

## Description of Automated Test 3

- Ensure tasks marked as deleted are not returned.
- Mark a task as deleted and verify it is not returned in the task list.

## Description of Automated Test 4

- Verify that tasks not belonging to the logged-in account are not returned.
- Create tasks for different accounts and ensure only the logged-in account's tasks are returned.

## Manual Test Cases Run:

**Test 1:** Created a task and shared it with multiple users via Postman. Verified that the task is shared correctly.

**Test 2:** Attempted to share a non-existent task and confirmed that an appropriate error message is returned.

## Description of Manual Test 1

- Create a task using Postman.
- Share the task with multiple users.
- Verify that the shared task is visible to those users.

## Description of Manual Test 2

- Attempt to share a task that does not exist using Postman.
- Verify that an error message is returned.

# Labels

- `feature`
- `task-sharing`
